docs/codeql/codeql-language-guides/customizing-library-models-for-ruby.rst

```diff
@@ -0,0 +1,406 @@
+        pack: codeql/ruby-all
+        extensible: sourceModel
+      data:
+        - [
+            "Sinatra::Base!",
+            "Method[get].Argument[block].Parameter[0]",
+            "remote",
+          ]
+
+- Since we're adding a new taint source, we add a tuple to the **sourceModel** extensible predicate.
+- The first column, **"Sinatra::Base!"**, begins the search at references to the **Sinatra::Base** class.
+  The **!** suffix indicates that we want to search for references to the class itself, rather than instances of the class.
+- **Method[get]** selects calls to the **get** method of the **Sinatra::Base** class.
+- **Argument[block]** selects the block argument to the **get** method call.
+- **Parameter[0]** selects the first parameter of the block argument (the parameter named **x**).
+- Finally, the kind **remote** indicates that this is considered a source of remote flow.
+
+Example: Using types to add MySQL injection sinks
+-------------------------------------------------
+
+In this example, we'll show how to add the following SQL injection sink:
+
+.. code-block:: ruby
+
+  def submit(q)
+    client = Mysql2::Client.new
+    client.query(q) # <-- add 'q' as a SQL injection sink
+  end
+
+We can recognize this using the following extension:
+
+.. code-block:: yaml
+
+  extensions:
+    - addsTo:
+        pack: codeql/ruby-all
+        extensible: sinkModel
+      data:
+        - ["Mysql2::Client", "Method[query].Argument[0]", "sql-injection"]
+
+- The first column, **"Mysql2::Client"**, begins the search at any instance of the **Mysql2::Client** class.
+- **Method[query]** selects any call to the **query** method on that instance.
+- **Argument[0]** selects the first argument to the method call.
+- **sql-injection** indicates that this is considered a sink for the SQL injection query.
+
+Continued example: Using type models
+------------------------------------
+
+Consider this variation on the previous example, the mysql2 EventMachine API is used.
+The client is obtained via a call to **Mysql2::EM::Client.new**.
+
+.. code-block:: ruby
+
+  def submit(client, q)
+    client = Mysql2::EM::Client.new
+    client.query(q)
+  end
+
+So far we have only one model for **Mysql2::Client**, but in the real world we
+may have many models for the various methods available. Because **Mysql2::EM::Client** is a subclass of **Mysql2::Client**, it inherits all of the same methods.
+Instead of updating all our models to include both classes, we can add a type
+model to indicate that **Mysql2::EM::Client** is a subclass of **Mysql2::Client**:
+
+.. code-block:: yaml
+
+  extensions:
+    - addsTo:
+        pack: codeql/ruby-all
+        extensible: typeModel
+      data:
+        - ["Mysql2::Client", "Mysql2::EM::Client", ""]
+
+Example: Adding flow through 'URI.decode_uri_component'
+-------------------------------------------------------
+
+In this example, we'll show how to add flow through calls to `URI.decode_uri_component`:
```

Comment:
```suggestion
In this example, we'll show how to add flow through calls to 'URI.decode_uri_component':
```

Another one where I think it'll otherwise render in italics, and would be ideal to use the same format throughout the doc